mypy/typeops.py

```diff
@@ -41,37 +41,49 @@ def type_object_type_from_function(signature: FunctionLike,
                                    def_info: TypeInfo,
                                    fallback: Instance,
                                    is_new: bool) -> FunctionLike:
-    # The __init__ method might come from a generic superclass
-    # (init_or_new.info) with type variables that do not map
-    # identically to the type variables of the class being constructed
-    # (info). For example
+    # We first need to record all non-trivial (explicit) self types in __init__,
+    # since they will not be available after we bind them. Note, we use explicit
+    # self-types only in the defining class, similar to __new__ (see below).
```

Comment:
`__new__` can use explicit returns in subclasses, but kind of as a technicality, since it only works using bound type vars, I think.

Reply:
Yes, I will make the comment more clear.